helix-rest/src/test/java/org/apache/helix/rest/metadatastore/TestZkMetadataStoreDirectory.java

```diff
@@ -344,6 +345,10 @@ public void testDataDeletionCallback() throws Exception {
               + " contains either empty or invalid routing data!")) {
             return false;
           }
+        } catch (IllegalArgumentException iae) {
+          if (!iae.getMessage().equals("Provided path is not a valid Zookeeper path: anyKey")) {
+            return false;
```

Comment:
The meaning of the block on line 343 is "ensure that an `IllegalStateException` is raised and the message is 'Failed to get...'". Is this block ensuring a `IllegalArgumentException` is raised? That's not what this test case is testing.

Reply:
Refreshing routing has a delay. And this is why `TestHelper.verify()` is used. Before routing is deleted in cached, namespace is still in cache, right? So the request still goes to TrieRouting, and IllegalArgumentException would be thrown because `anyKey` is invalid. In this case, we should catch the exception and retry, right?

The better I think is this. I will update PR.
```
    Assert.assertTrue(TestHelper.verify(() -> {
      for (String namespace : _routingZkAddrMap.keySet()) {
        try {
          _metadataStoreDirectory.getMetadataStoreRealm(namespace, "anyKey");
          Assert.fail("Should not successfully get routing data");
        } catch (IllegalStateException e) {
          // If other IllegalStateException, it is unexpected and this test should fail.
          if (!e.getMessage().equals("Failed to get metadata store realm: Namespace " + namespace
              + " contains either empty or invalid routing data!")) {
            throw e;
          }
        } catch (IllegalArgumentException iae) {
          // If routing data is not yet refreshed, return false and retry.
          if (iae.getMessage().equals("Provided path is not a valid Zookeeper path: anyKey")) {
            return false;
          }
          // If other IllegalArgumentException, it is not expected and this test should fail.
          throw iae;
        }
      }
      return true;
    }, TestHelper.WAIT_DURATION));
```

Reply:
I see your point now, thanks for the explanation. Was this test case flaky because of that? This is a very good catch! 👍  
I think it makes more sense to replace "anyKey" with an actual existing key in the routing data; the handling of `IllegalArgumentException` is only caused by "anyKey", and is not what this test case should focus on. My point is that using "anyKey" is a mistake. We shouldn't change the test case to accommodate that mistake; we can just correct it by using something like `TEST_SHARDING_KEYS`.